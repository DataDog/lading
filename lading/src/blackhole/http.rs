//! The HTTP protocol speaking blackhole.
//!
//! ## Metrics
//!
//! `bytes_received`: Total bytes received
//! `requests_received`: Total requests received
//!

use std::{net::SocketAddr, time::Duration};

use http::{header::InvalidHeaderValue, status::InvalidStatusCode, HeaderMap};
use hyper::{
    body::HttpBody,
    header,
    server::conn::{AddrIncoming, AddrStream},
    service::{make_service_fn, service_fn},
    Body, Request, Response, Server, StatusCode,
};
use metrics::{counter, Counter};
use serde::{Deserialize, Serialize};
use tower::ServiceBuilder;
use tracing::{debug, error, info};

use super::General;

fn default_concurrent_requests_max() -> usize {
    100
}

/// Errors produced by [`Http`].
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// Wrapper for [`hyper::Error`].
    #[error("HTTP server error: {0}")]
    Hyper(hyper::Error),
    /// The configured content type value was not valid.
    #[error("The configured content type value was not valid: {0}")]
    InvalidContentType(InvalidHeaderValue),
    /// The configured status code was not valid.
    #[error("The configured status code was not valid: {0}")]
    InvalidStatusCode(InvalidStatusCode),
    /// Failed to deserialize the configuration.
    #[error("Failed to deserialize the configuration: {0}")]
    Serde(#[from] serde_json::Error),
}

/// Body variant supported by this blackhole.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub enum BodyVariant {
    /// All response bodies will be empty.
    Nothing,
    /// All response bodies will mimic AWS Kinesis.
    AwsKinesis,
    /// Respond with a hardcoded byte slice value
    RawBytes,
    /// Respond with a hardcoded string value
    Static(String),
}

fn default_body_variant() -> BodyVariant {
    BodyVariant::Nothing
}

fn default_response_delay_millis() -> u64 {
    0
}

fn default_status_code() -> u16 {
    StatusCode::OK.as_u16()
}

fn default_headers() -> HeaderMap {
    let mut map = HeaderMap::new();
    map.insert(
        header::CONTENT_TYPE,
        "application/json"
            .parse()
            .expect("Not possible to parse into HeaderMap"),
    );
    map
}

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
/// Configuration for [`Http`]
pub struct Config {
    /// number of concurrent HTTP connections to allow
    #[serde(default = "default_concurrent_requests_max")]
    pub concurrent_requests_max: usize,
    /// address -- IP plus port -- to bind to
    pub binding_addr: SocketAddr,
    /// the body variant to respond with, default nothing
    #[serde(default = "default_body_variant")]
    pub body_variant: BodyVariant,
    /// Headers to include in the response; default is `Content-Type: application/json`
    #[serde(with = "http_serde::header_map", default = "default_headers")]
    pub headers: HeaderMap,
    /// the content-type header to respond with, defaults to 200
    #[serde(default = "default_status_code")]
    pub status: u16,
    /// raw array of bytes if the `raw_bytes` body variant is selected
    #[serde(default)]
    pub raw_bytes: Vec<u8>,
    /// delay to add before making a response
    #[serde(default = "default_response_delay_millis")]
    pub response_delay_millis: u64,
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case")]
struct KinesisPutRecordBatchResponseEntry {
    error_code: Option<String>,
    error_message: Option<String>,
    record_id: String,
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case")]
struct KinesisPutRecordBatchResponse {
    encrypted: Option<bool>,
    failed_put_count: u32,
    request_responses: Vec<KinesisPutRecordBatchResponseEntry>,
}

#[allow(clippy::borrow_interior_mutable_const)]
async fn srv(
    status: StatusCode,
    bytes_received: Counter,
    requests_received: Counter,
    body_bytes: Vec<u8>,
    req: Request<Body>,
    headers: HeaderMap,
    response_delay: Duration,
) -> Result<Response<Body>, hyper::Error> {
    requests_received.increment(1);

    let (parts, body) = req.into_parts();

    let bytes = body.collect().await?.to_bytes();

    match crate::codec::decode(parts.headers.get(hyper::header::CONTENT_ENCODING), bytes) {
        Err(response) => Ok(response),
        Ok(body) => {
            bytes_received.increment(body.len() as u64);

            tokio::time::sleep(response_delay).await;

            let mut okay = Response::default();
            *okay.status_mut() = status;
            *okay.headers_mut() = headers;
            *okay.body_mut() = Body::from(body_bytes);
            Ok(okay)
        }
    }
}

#[derive(Debug)]
/// The HTTP blackhole.
pub struct Http {
    httpd_addr: SocketAddr,
    body_bytes: Vec<u8>,
    concurrency_limit: usize,
    shutdown: lading_signal::Watcher,
    headers: HeaderMap,
    status: StatusCode,
    metric_labels: Vec<(String, String)>,
    response_delay: Duration,
}

impl Http {
    /// Create a new [`Http`] server instance
    ///
    /// # Errors
    ///
    /// Returns an error if the configuration is invalid.
    ///
    /// # Panics
    ///
    /// None known.
    pub fn new(
        general: General,
        config: &Config,
        shutdown: lading_signal::Watcher,
    ) -> Result<Self, Error> {
        let status = StatusCode::from_u16(config.status).map_err(Error::InvalidStatusCode)?;

        let mut metric_labels = vec![
            ("component".to_string(), "blackhole".to_string()),
            ("component_name".to_string(), "http".to_string()),
        ];
        if let Some(id) = general.id {
            metric_labels.push(("id".to_string(), id));
        }

        let body_bytes = match &config.body_variant {
            BodyVariant::AwsKinesis => {
                let response = KinesisPutRecordBatchResponse {
                    encrypted: None,
                    failed_put_count: 0,
                    request_responses: vec![KinesisPutRecordBatchResponseEntry {
                        error_code: None,
                        error_message: None,
                        record_id: "foobar".to_string(),
                    }],
                };
                serde_json::to_vec(&response)?
            }
            BodyVariant::Nothing => vec![],
            BodyVariant::RawBytes => config.raw_bytes.clone(),
            BodyVariant::Static(val) => val.as_bytes().to_vec(),
        };

        Ok(Self {
            httpd_addr: config.binding_addr,
            body_bytes,
            concurrency_limit: config.concurrent_requests_max,
            headers: config.headers.clone(),
            status,
            shutdown,
            metric_labels,
            response_delay: Duration::from_millis(config.response_delay_millis),
        })
    }

    /// Run [`Http`] to completion
    ///
    /// This function runs the HTTP server forever, unless a shutdown signal is
    /// received or an unrecoverable error is encountered.
    ///
    /// # Errors
    ///
    /// Function will return an error if the configuration is invalid or if
    /// receiving a packet fails.
    pub async fn run(self) -> Result<(), Error> {
        let bytes_received = counter!("bytes_received", &self.metric_labels);
        let requests_received = counter!("requests_received", &self.metric_labels);
        let service = make_service_fn(|_: &AddrStream| {
            let bytes_received = bytes_received.clone();
            let requests_received = requests_received.clone();
            let body_bytes = self.body_bytes.clone();
            let headers = self.headers.clone();
            async move {
                Ok::<_, hyper::Error>(service_fn(move |request| {
                    debug!("REQUEST: {:?}", request);
                    srv(
                        self.status,
                        bytes_received.clone(),
                        requests_received.clone(),
                        body_bytes.clone(),
                        request,
                        headers.clone(),
                        self.response_delay,
                    )
                }))
            }
        });
        let svc = ServiceBuilder::new()
            .load_shed()
            .concurrency_limit(self.concurrency_limit)
            .timeout(Duration::from_secs(1))
            .service(service);

        let addr = AddrIncoming::bind(&self.httpd_addr)
            .map(|mut addr| {
                addr.set_keepalive(Some(Duration::from_secs(60)));
                addr
            })
            .map_err(Error::Hyper)?;

        let server = Server::builder(addr).serve(svc);
        tokio::select! {
            res = server => {
                error!("server shutdown unexpectedly");
                 res.map_err(Error::Hyper)
            }
            () = self.shutdown.recv() => {
                info!("shutdown signal received");
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn config_deserializes_variant_nothing() {
        let contents = r#"
binding_addr: "127.0.0.1:1000"
body_variant: "nothing"
"#;
        let config: Config =
            serde_yaml::from_str(contents).expect("Contents do not match the structure expected");
        assert_eq!(
            config,
            Config {
                concurrent_requests_max: default_concurrent_requests_max(),
                response_delay_millis: default_response_delay_millis(),
                binding_addr: SocketAddr::from_str("127.0.0.1:1000")
                    .expect("Not possible to parse into SocketAddr"),
                body_variant: BodyVariant::Nothing,
                headers: default_headers(),
                status: default_status_code(),
                raw_bytes: vec![],
            },
        );
    }

    #[test]
    fn config_deserializes_raw_bytes() {
        let contents = r#"
binding_addr: "127.0.0.1:1000"
body_variant: "raw_bytes"
raw_bytes: [0x01, 0x02, 0x10]
"#;
        let config: Config =
            serde_yaml::from_str(contents).expect("Contents do not match the structure expected");
        assert_eq!(
            config,
            Config {
                concurrent_requests_max: default_concurrent_requests_max(),
                response_delay_millis: default_response_delay_millis(),
                binding_addr: SocketAddr::from_str("127.0.0.1:1000")
                    .expect("Not possible to parse into SocketAddr"),
                body_variant: BodyVariant::RawBytes,
                headers: default_headers(),
                status: default_status_code(),
                raw_bytes: vec![0x01, 0x02, 0x10],
            },
        );
    }
}
