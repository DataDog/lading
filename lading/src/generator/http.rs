//! The HTTP protocol speaking generator.
//!
//! ## Metrics
//!
//! `requests_sent`: Total number of requests sent
//! `request_ok`: Successful requests
//! `request_failure`: Failed requests
//! `bytes_written`: Total bytes written
//! `bytes_per_second`: Configured rate to send data
//!
//! Additional metrics may be emitted by this generator's [throttle].
//!

use std::{num::NonZeroU32, thread};

use byte_unit::ByteError;
use hyper::{
    client::{Client, HttpConnector},
    header::CONTENT_LENGTH,
    Body, HeaderMap, Request, Uri,
};
use lading_throttle::Throttle;
use metrics::{counter, gauge};
use once_cell::sync::OnceCell;
use rand::{prelude::StdRng, SeedableRng};
use serde::{Deserialize, Serialize};
use tokio::sync::{mpsc, Semaphore};
use tracing::info;

use crate::common::PeekableReceiver;
use lading_payload::block::{self, Block};

use super::General;

static CONNECTION_SEMAPHORE: OnceCell<Semaphore> = OnceCell::new();

/// The HTTP method to be used in requests
#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "snake_case")]
#[serde(deny_unknown_fields)]
pub enum Method {
    /// Make HTTP Post requests
    Post {
        /// The payload generator to use for this target
        variant: lading_payload::Config,
        /// The maximum size in bytes of the cache of prebuilt messages
        maximum_prebuild_cache_size_bytes: byte_unit::Byte,
        /// Whether to use a fixed or streaming block cache
        #[serde(default = "lading_payload::block::default_cache_method")]
        block_cache_method: block::CacheMethod,
    },
}

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
/// Configuration of this generator.
pub struct Config {
    /// The seed for random operations against this target
    pub seed: [u8; 32],
    /// The URI for the target, must be a valid URI
    #[serde(with = "http_serde::uri")]
    pub target_uri: Uri,
    /// The method to use against the URI
    pub method: Method,
    /// Headers to include in the request
    #[serde(with = "http_serde::header_map")]
    pub headers: HeaderMap,
    /// The bytes per second to send or receive from the target
    pub bytes_per_second: byte_unit::Byte,
    /// The maximum size in bytes of the largest block in the prebuild cache.
    #[serde(default = "lading_payload::block::default_maximum_block_size")]
    pub maximum_block_size: byte_unit::Byte,
    /// The total number of parallel connections to maintain
    pub parallel_connections: u16,
    /// The load throttle configuration
    #[serde(default)]
    pub throttle: lading_throttle::Config,
}

#[derive(thiserror::Error, Debug)]
/// Errors produced by [`Http`].
pub enum Error {
    /// Wrapper around [`std::io::Error`].
    #[error("Io error: {0}")]
    Io(#[from] ::std::io::Error),
    /// Creation of payload blocks failed.
    #[error("Block creation error: {0}")]
    Block(#[from] block::Error),
    /// Wrapper around [`hyper::Error`].
    #[error("Hyper error: {0}")]
    Hyper(#[from] hyper::Error),
    /// Wrapper around [`hyper::http::Error`].
    #[error("HTTP error: {0}")]
    Http(#[from] hyper::http::Error),
    //  OTLP traffic
    //#[error("OTLP error: {0}")]
    //Otlp(#[from] OtlpError),
    /// Byte error
    #[error("Bytes must not be negative: {0}")]
    Byte(#[from] ByteError),
    /// Failed to convert, value is 0
    #[error("Value provided must not be zero")]
    Zero,
}

/// The HTTP generator.
///
/// This generator is reposnsible for connecting to the target via HTTP. Today
/// only POST and GET are supported.
#[derive(Debug)]
pub struct Http {
    uri: Uri,
    method: hyper::Method,
    headers: hyper::HeaderMap,
    parallel_connections: u16,
    throttle: Throttle,
    block_cache: block::Cache,
    metric_labels: Vec<(String, String)>,
    shutdown: lading_signal::Watcher,
}

impl Http {
    /// Create a new [`Http`] instance
    ///
    /// # Errors
    ///
    /// Creation will fail if the underlying governor capacity exceeds u32.
    ///
    /// # Panics
    ///
    /// Function will panic if user has passed non-zero values for any byte
    /// values. Sharp corners.
    #[allow(clippy::cast_possible_truncation)]
    pub fn new(
        general: General,
        config: Config,
        shutdown: lading_signal::Watcher,
    ) -> Result<Self, Error> {
        let mut rng = StdRng::from_seed(config.seed);
        let mut labels = vec![
            ("component".to_string(), "generator".to_string()),
            ("component_name".to_string(), "http".to_string()),
        ];
        if let Some(id) = general.id {
            labels.push(("id".to_string(), id));
        }

        let bytes_per_second = NonZeroU32::new(config.bytes_per_second.get_bytes() as u32)
            .expect("config bytes per second must be non-zero");
        gauge!("bytes_per_second", &labels).set(f64::from(bytes_per_second.get()));

        match config.method {
            Method::Post {
                variant,
                maximum_prebuild_cache_size_bytes,
                block_cache_method,
            } => {
                let total_bytes =
                    NonZeroU32::new(maximum_prebuild_cache_size_bytes.get_bytes() as u32)
                        .ok_or(Error::Zero)?;
                let block_cache = match block_cache_method {
                    block::CacheMethod::Fixed => block::Cache::fixed(
                        &mut rng,
                        total_bytes,
                        config.maximum_block_size.get_bytes(),
                        &variant,
                    )?,
                };

                CONNECTION_SEMAPHORE
                    .set(Semaphore::new(config.parallel_connections as usize))
                    .expect("failed to set semaphore");

                Ok(Self {
                    parallel_connections: config.parallel_connections,
                    uri: config.target_uri,
                    method: hyper::Method::POST,
                    headers: config.headers,
                    block_cache,
                    throttle: Throttle::new_with_config(config.throttle, bytes_per_second),
                    metric_labels: labels,
                    shutdown,
                })
            }
        }
    }

    /// Run [`Http`] to completion or until a shutdown signal is received.
    ///
    /// # Errors
    ///
    /// TODO
    ///
    /// # Panics
    ///
    /// Function will panic if it is unable to create HTTP requests for the
    /// target.
    pub async fn spin(mut self) -> Result<(), Error> {
        let client: Client<HttpConnector, Body> = Client::builder()
            .pool_max_idle_per_host(self.parallel_connections as usize)
            .retry_canceled_requests(false)
            .build_http();
        let method = self.method;
        let uri = self.uri;

        let labels = self.metric_labels;
        // Move the block_cache into an OS thread, exposing a channel between it
        // and this async context.
        let block_cache = self.block_cache;
        let (snd, rcv) = mpsc::channel(1024);
        let mut rcv: PeekableReceiver<Block> = PeekableReceiver::new(rcv);
        thread::Builder::new().spawn(|| block_cache.spin(snd))?;

        let shutdown_wait = self.shutdown.recv();
        tokio::pin!(shutdown_wait);
        loop {
            let blk = rcv.next().await.expect("block cache should never be empty");
            let total_bytes = blk.total_bytes;

            let body = Body::from(blk.bytes.clone());
            let block_length = blk.bytes.len();

            let mut request: Request<Body> = Request::builder()
                .method(method.clone())
                .uri(&uri)
                .header(CONTENT_LENGTH, block_length)
                .body(body)?;
            let headers = request.headers_mut();
            for (k, v) in self.headers.clone().drain() {
                if let Some(k) = k {
                    headers.insert(k, v);
                }
            }

            tokio::select! {
                _ = self.throttle.wait_for(total_bytes) => {
                    let client = client.clone();
                    let labels = labels.clone();

                    let permit = CONNECTION_SEMAPHORE.get().expect("Connection Semaphore is being initialized or cell is empty").acquire().await.expect("Connection Semaphore has already closed");
                    tokio::spawn(async move {
                        counter!("requests_sent", &labels).increment(1);
                        match client.request(request).await {
                            Ok(response) => {
                                counter!("bytes_written", &labels).increment(block_length as u64);
                                let status = response.status();
                                let mut status_labels = labels.clone();
                                status_labels
                                    .push(("status_code".to_string(), status.as_u16().to_string()));
                                counter!("request_ok", &status_labels).increment(1);
                            }
                            Err(err) => {
                                let mut error_labels = labels.clone();
                                error_labels.push(("error".to_string(), err.to_string()));
                                counter!("request_failure", &error_labels).increment(1);
                            }
                        }
                        drop(permit);
                    });
                },
                () = &mut shutdown_wait => {
                    info!("shutdown signal received");
                    // Acquire all available connections, meaning that we have
                    // no outstanding tasks in flight.
                    let _semaphore = CONNECTION_SEMAPHORE.get().expect("Connection Semaphore is being initialized or cell is empty").acquire_many(u32::from(self.parallel_connections)).await.expect("Connection Semaphore has already closed");
                    return Ok(());
                },
            }
        }
    }
}
