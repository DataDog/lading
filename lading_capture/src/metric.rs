use metrics::Key;
use std::time::Instant;

#[derive(Debug, Clone)]
pub(crate) enum CounterValue {
    Increment(u64),
    Absolute(u64),
}

#[derive(Debug, Clone)]
pub(crate) enum GaugeValue {
    Increment(f64),
    Decrement(f64),
    Set(f64),
}

#[derive(Debug, Clone)]
pub(crate) struct Counter {
    pub key: Key,
    pub timestamp: Instant,
    pub value: CounterValue,
}

#[derive(Debug, Clone)]
pub(crate) struct Gauge {
    pub key: Key,
    pub timestamp: Instant,
    pub value: GaugeValue,
}

#[derive(Debug, Clone)]
pub(crate) enum Metric {
    /// Counter increment
    Counter(Counter),
    /// Gauge set
    Gauge(Gauge),
}
