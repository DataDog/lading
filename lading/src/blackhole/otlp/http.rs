//! HTTP implementation of the OTLP blackhole.

use bytes::Bytes;
use http::StatusCode;
use http_body_util::{BodyExt, combinators::BoxBody};
use hyper::{Request, Response};
use metrics::counter;
use opentelemetry_proto::tonic::collector::logs::v1::{
    ExportLogsServiceRequest, ExportLogsServiceResponse,
};
use opentelemetry_proto::tonic::collector::metrics::v1::{
    ExportMetricsServiceRequest, ExportMetricsServiceResponse,
};
use opentelemetry_proto::tonic::collector::trace::v1::{
    ExportTraceServiceRequest, ExportTraceServiceResponse,
};
use opentelemetry_proto::tonic::metrics::v1::metric::Data as MetricData;
use prost::Message;
use std::net::SocketAddr;
use std::time::Duration;
use tokio::task::JoinHandle;
use tracing::{error, info};

use crate::blackhole::common;

/// Run the HTTP server for OTLP
pub(crate) fn run_server(
    addr: SocketAddr,
    response_delay: Duration,
    base_labels: &[(String, String)],
    concurrency_limit: usize,
    shutdown: lading_signal::Watcher,
) -> JoinHandle<()> {
    info!("Starting OTLP HTTP server on {addr}");
    let mut labels = Vec::with_capacity(base_labels.len() + 1);
    labels.push(("protocol".to_string(), "http".to_string()));
    labels.extend_from_slice(base_labels);

    tokio::spawn(async move {
        let handler = OtlpHttpHandler::new(response_delay, &labels);

        let result = common::run_httpd(addr, concurrency_limit, shutdown, labels, move || {
            let handler_clone = handler.clone();
            hyper::service::service_fn(move |req| {
                let request_handler = handler_clone.clone();
                request_handler.handle_request(req)
            })
        })
        .await;

        if let Err(e) = result {
            error!("HTTP server error: {e}");
        }
    })
}

/// Handler for OTLP HTTP requests
#[derive(Clone, Debug)]
struct OtlpHttpHandler {
    labels: Vec<(String, String)>,
    empty_metrics_response: Bytes,
    empty_traces_response: Bytes,
    empty_logs_response: Bytes,
    content_type_header_value: http::HeaderValue,
    response_delay: Duration,
}

impl OtlpHttpHandler {
    fn new(response_delay: Duration, labels: &[(String, String)]) -> Self {
        // Pre-compute empty responses
        let empty_metrics_response =
            Bytes::from(ExportMetricsServiceResponse::default().encode_to_vec());
        let empty_traces_response =
            Bytes::from(ExportTraceServiceResponse::default().encode_to_vec());
        let empty_logs_response = Bytes::from(ExportLogsServiceResponse::default().encode_to_vec());

        let content_type_header_value = "application/x-protobuf"
            .parse()
            .expect("application/x-protobuf is a valid MIME type");

        Self {
            labels: labels.to_vec(),
            empty_metrics_response,
            empty_traces_response,
            empty_logs_response,
            content_type_header_value,
            response_delay,
        }
    }

    #[inline]
    async fn build_response(
        &self,
        response_bytes: Bytes,
    ) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
        if self.response_delay.as_micros() > 0 {
            tokio::time::sleep(self.response_delay).await;
        }

        let mut response = Response::builder().status(StatusCode::OK);
        let headers = response
            .headers_mut()
            .expect("Response builder should always provide headers_mut");
        headers.insert(
            hyper::header::CONTENT_TYPE,
            self.content_type_header_value.clone(),
        );
        Ok(response
            .body(crate::full(response_bytes))
            .expect("Creating HTTP response should not fail"))
    }

    async fn handle_request(
        self,
        req: Request<hyper::body::Incoming>,
    ) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
        // Fast-path for invalid paths
        let path_ref = req.uri().path();
        if path_ref != "/v1/metrics" && path_ref != "/v1/traces" && path_ref != "/v1/logs" {
            static NOT_FOUND: &[u8] = b"Not found";
            return Ok(Response::builder()
                .status(StatusCode::NOT_FOUND)
                .body(crate::full(Bytes::from_static(NOT_FOUND)))
                .expect("Creating HTTP response should not fail"));
        }

        counter!("requests_received", &self.labels).increment(1);

        // Check for empty bodies using Content-Length when available
        if let Some(content_length) = req.headers().get(hyper::header::CONTENT_LENGTH) {
            if let Ok(length) = content_length.to_str() {
                if let Ok(length) = length.parse::<u64>() {
                    if length == 0 {
                        counter!("bytes_received", &self.labels).increment(0);

                        let response_bytes = match path_ref {
                            "/v1/metrics" => self.empty_metrics_response.clone(),
                            "/v1/traces" => self.empty_traces_response.clone(),
                            "/v1/logs" => self.empty_logs_response.clone(),
                            _ => unreachable!(), // checked earlier
                        };

                        return self.build_response(response_bytes).await;
                    }
                }
            }
        }

        let content_encoding = req.headers().get(hyper::header::CONTENT_ENCODING).cloned();
        let path = path_ref.to_string();
        let (_, body) = req.into_parts();

        let body_bytes = body.collect().await?.to_bytes();

        counter!("bytes_received", &self.labels).increment(body_bytes.len() as u64);
        let response_bytes =
            match crate::codec::decode(content_encoding.as_ref(), body_bytes.clone()) {
                Ok(decoded) => {
                    counter!("decoded_bytes_received", &self.labels)
                        .increment(decoded.len() as u64);

                    match path.as_str() {
                        "/v1/metrics" => self.process_metrics(&decoded),
                        "/v1/traces" => self.process_traces(&decoded),
                        "/v1/logs" => self.process_logs(&decoded),
                        _ => unreachable!(
                            "path previously checked, catastrophic programming mistake"
                        ),
                    }
                }
                Err(response) => return Ok(response),
            };

        self.build_response(response_bytes).await
    }

    fn process_metrics(&self, body_bytes: &[u8]) -> Bytes {
        if body_bytes.is_empty() {
            return self.empty_metrics_response.clone();
        }

        let mut total_points: u64 = 0;
        if let Ok(request) = ExportMetricsServiceRequest::decode(body_bytes) {
            for rm in &request.resource_metrics {
                for sm in &rm.scope_metrics {
                    for m in &sm.metrics {
                        if let Some(data) = &m.data {
                            let points_count = match data {
                                MetricData::Gauge(g) => g.data_points.len(),
                                MetricData::Sum(s) => s.data_points.len(),
                                MetricData::Histogram(h) => h.data_points.len(),
                                MetricData::ExponentialHistogram(eh) => eh.data_points.len(),
                                MetricData::Summary(s) => s.data_points.len(),
                            };
                            total_points += points_count as u64;
                        }
                    }
                }
            }
        }

        if total_points > 0 {
            counter!("data_points_received", &self.labels).increment(total_points);
        }

        self.empty_metrics_response.clone()
    }

    fn process_traces(&self, body_bytes: &[u8]) -> Bytes {
        if body_bytes.is_empty() {
            return self.empty_traces_response.clone();
        }

        let mut total_spans: u64 = 0;
        if let Ok(request) = ExportTraceServiceRequest::decode(body_bytes) {
            for rs in &request.resource_spans {
                for ss in &rs.scope_spans {
                    let spans_count = ss.spans.len() as u64;
                    total_spans += spans_count;
                }
            }
        }

        if total_spans > 0 {
            counter!("data_points_received", &self.labels).increment(total_spans);
        }

        self.empty_traces_response.clone()
    }

    fn process_logs(&self, body_bytes: &[u8]) -> Bytes {
        if body_bytes.is_empty() {
            return self.empty_logs_response.clone();
        }

        let mut total_logs: u64 = 0;
        if let Ok(request) = ExportLogsServiceRequest::decode(body_bytes) {
            for rl in &request.resource_logs {
                for sl in &rl.scope_logs {
                    let logs_count = sl.log_records.len() as u64;
                    total_logs += logs_count;
                }
            }
        }

        if total_logs > 0 {
            counter!("data_points_received", &self.labels).increment(total_logs);
        }

        self.empty_logs_response.clone()
    }
}
