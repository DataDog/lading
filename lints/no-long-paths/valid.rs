// Cases that should NOT trigger the lint. If custom_lints triggers in this file
// the tool is NOT working as intended.

// Use statements should be excluded (even with 3+ segments)
use self::module::submodule::function;
use crate::generator::http::Server;
use foo::{bar::baz, bard::baz::bing, bing::foo};
use lading_payload::dogstatsd::metric::Counter;
use std::collections::{BTreeMap, HashMap};
use tokio::sync::broadcast::Sender;

// Two segments are fine
fn example_two_segments() {
    let config = Config::new();
    let server = Server::start();
    let file = File::open("path");
    let result = Response::ok();
}

// Comments and strings should not trigger
fn strings_and_comments() {
    // This comment mentions some::long::path::here but shouldn't trigger
    let string = "some::long::path::in::string";
    let raw_string = r#"another::long::path::here"#;
}

// Simple macros
fn macro_cases() {
    println!("valid");
    format!("string with {}", some_variable);
}
