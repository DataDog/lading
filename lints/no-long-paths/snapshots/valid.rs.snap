error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
  ┌─ lints/no-long-paths/valid.rs:5:16
  │
5 │     let file = fs::File::open("path");
  │                ^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
  ┌─ lints/no-long-paths/valid.rs:6:18
  │
6 │     let result = http::Response::new();
  │                  ^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
  ┌─ lints/no-long-paths/valid.rs:7:16
  │
7 │     let data = json::Value::from(42);
  │                ^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:12:5
   │
12 │ use tokio::sync::broadcast::Sender;
   │     ^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:13:5
   │
13 │ use crate::generator::http::Server;
   │     ^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:14:5
   │
14 │ use lading_payload::dogstatsd::metric::Counter;
   │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:18:16
   │
18 │     let path = std::env::current_dir().unwrap();
   │                ^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:19:16
   │
19 │     let file = std::fs::File::create("test.txt").unwrap();
   │                ^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:19:16
   │
19 │     let file = std::fs::File::create("test.txt").unwrap();
   │                ^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:20:18
   │
20 │     let thread = std::thread::Builder::new().spawn(|| {});
   │                  ^^^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:20:18
   │
20 │     let thread = std::thread::Builder::new().spawn(|| {});
   │                  ^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:25:14
   │
25 │     let rt = tokio::runtime::Builder::new_multi_thread().build();
   │              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:25:14
   │
25 │     let rt = tokio::runtime::Builder::new_multi_thread().build();
   │              ^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:26:16
   │
26 │     let file = tokio::fs::File::create("test.txt").await;
   │                ^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:26:16
   │
26 │     let file = tokio::fs::File::create("test.txt").await;
   │                ^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:27:18
   │
27 │     let sender = tokio::sync::broadcast::channel(10);
   │                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[no-long-paths]: Found long path (3+ segments) - import a subset of this path
   ┌─ lints/no-long-paths/valid.rs:27:18
   │
27 │     let sender = tokio::sync::broadcast::channel(10);
   │                  ^^^^^^^^^^^^^^^^^^^^^^

Error: 17 error(s) found in code.
Help: Scan succeeded and found error level diagnostics in the codebase.
